### PR TITLE
LSP error flusher

### PR DESCRIPTION
**Pull request by vilhena-stripe**
- Adds an error flusher for LSP
- Change `pipeline::typecheck` so that instead of marking files for flushing, we flush per file after worker finishes typechecking each file
- Add "flusher" for autocorrects that stores the errors for later access
- Move critical error checks and flushing from `pipeline.cc` to `realmain.cc` so they only trigger for the CLI path

### Motivation
Part of project for sorbet fast feedback to allow streaming errors per file

### Test plan
Existing tests should cover

### CODE REVIEW

File: core/ErrorQueue.cc [hunk 1]
```diff
@@ -87,20 +83,41 @@ vector<unique_ptr<core::Error>> ErrorQueue::drainAllErrors() {
...
+    Timer timeit(tracer, "ErrorQueue::flushErrorsForFile");
+
+    core::ErrorQueueMessage msg;
+    for (auto result = queue.try_pop(msg); result.gotItem(); result = queue.try_pop(msg)) {
+        collected[msg.whatFile].emplace_back(move(msg));
+    }
+
+    vector<unique_ptr<core::ErrorQueueMessage>> errors;
+    for (auto &error : collected[file]) {
+        errors.emplace_back(make_unique<core::ErrorQueueMessage>(move(error)));
```
Comment (jvilk-stripe) [inline core/ErrorQueue.cc:116]: qq: Can we change `collected` to be `UnorderedMap<core::FileRef, std::vector<std::unique_ptr<ErrorQueueMessage>>`? Then we can replace these loops with `move(collected[file])`.

File: core/ErrorFlusherStdout.h [hunk 2]
```diff
@@ -17,7 +17,8 @@ class ErrorFlusherStdout : public ErrorFlusher {
     ErrorFlusherStdout() = default;
     ~ErrorFlusherStdout() = default;
 
-    void flushErrors(spdlog::logger &logger, std::vector<std::unique_ptr<ErrorQueueMessage>> error) override;
+    void flushErrors(spdlog::logger &logger, std::vector<std::unique_ptr<ErrorQueueMessage>> errors,
+                     const GlobalState &gs, core::FileRef file) override;
     void flushErrorCount(spdlog::logger &logger, int count) override;
```
Comment (jvilk-stripe) [inline core/ErrorFlusherStdout.h:22]: `flushErrorCount` and `flushAutocorrects` are only relevant on the CLI. Can we remove them from the interface and `ErrorQueue`, and have `realmain` call them directly on `ErrorFlusherStdout`?

Comment (jvilk-stripe) [reply core/ErrorFlusherStdout.h:22]: (`ErrorFlusherStdout` can simply count the number of errors it prints, so we can remove the atomic error counter from `ErrorQueue`)

File: main/lsp/AutocorrectFlusher.cc [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#include "main/lsp/AutocorrectFlusher.h"
+
+using namespace std;
+namespace sorbet::realmain::lsp {
+
+void AutocorrectFlusher::flushErrors(spdlog::logger &logger, vector<unique_ptr<core::ErrorQueueMessage>> errors,
+                                     const core::GlobalState &gs, core::FileRef file) {
+    for (auto &error : errors) {
+        if (error->error->isSilenced) {
```
Comment (jvilk-stripe) [inline main/lsp/AutocorrectFlusher.cc:9]: You need to check that `error->kind` is `Error`, otherwise this may cause a bug. (While this should not actually happen, it could theoretically be a `Flush` or a `Query`, and it's best to program defensively.)

File: main/lsp/AutocorrectFlusher.h [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#ifndef SORBET_AUTOCORRECT_FLUSHER_H
+#define SORBET_AUTOCORRECT_FLUSHER_H
+
+#include "core/ErrorFlusher.h"
+
+namespace sorbet::realmain::lsp {
+
+class AutocorrectFlusher : public core::ErrorFlusher {
```
Comment (jvilk-stripe) [inline main/lsp/AutocorrectFlusher.h:8]: This flusher is only used within `LSPTypechecker.cc`, correct? If so, we can define it there and not have a header file at all.

File: main/lsp/ErrorFlusherLSP.cc [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+namespace sorbet::realmain::lsp {
+ErrorFlusherLSP::ErrorFlusherLSP(const u4 epoch, shared_ptr<ErrorReporter> errorReporter)
+    : epoch(epoch), errorReporter(errorReporter){};
+
+void ErrorFlusherLSP::flushErrors(spdlog::logger &logger, vector<unique_ptr<core::ErrorQueueMessage>> errors,
+                                  const core::GlobalState &gs, core::FileRef file) {
+    vector<std::unique_ptr<core::Error>> errorsAccumulated;
+
+    for (auto &error : errors) {
+        if (error->error->isSilenced) {
```
Comment (jvilk-stripe) [inline main/lsp/ErrorFlusherLSP.cc:13]: Similarly, you need to verify that this is a message of kind `Error`

File: main/lsp/ErrorReporter.cc [hunk 6]
```diff
@@ -22,7 +22,9 @@ vector<core::FileRef> ErrorReporter::filesWithErrorsSince(u4 epoch) {
 void ErrorReporter::beginEpoch(u4 epoch, vector<unique_ptr<Timer>> diagnosticLatencyTimers) {
     ENFORCE(epochTimers.find(epoch) == epochTimers.end());
     vector<Timer> firstDiagnosticLatencyTimers;
-
+    if (config->getClientConfig().enableTypecheckInfo) {
+        Timer::timedSleep(chrono::milliseconds(50), *config->logger, "delay so timer is reported");
```
Comment (jvilk-stripe) [inline main/lsp/ErrorReporter.cc:26]: `50` is a lot; does it work fine with `5`?

File: main/lsp/LSPTypechecker.cc [hunk 7]
```diff
@@ -415,16 +415,14 @@ bool LSPTypechecker::runSlowPath(LSPFileUpdates updates, WorkerPool &workers, bo
     // Note: This is important to do even if the slow path was canceled. It clears out any typechecking errors from the
     // aborted typechecking run.
     // Note 2: `gs` now holds the value of `finalGS`.
-    auto out = gs->errorQueue->drainWithQueryResponses();
+    gs->errorQueue->drainWithQueryResponses();
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.cc:418]: Since we create a brand new error queue for every typecheck run, I think we can get rid of this now with no consequences.

File: main/lsp/LSPTypechecker.cc [hunk 8]
```diff
@@ -498,12 +475,6 @@ void LSPTypechecker::commitFileUpdates(LSPFileUpdates &updates, bool couldBeCanc
     }
 }
 
-void LSPTypechecker::commitTypecheckRun(TypecheckRun run) {
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.cc:501]: 🎉

File: main/lsp/LSPTypechecker.h [hunk 9]
```diff
@@ -25,24 +26,6 @@ struct LSPQueryResult {
     std::unique_ptr<ResponseError> error;
 };
 
-class TypecheckRun final {
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.h:28]: 🎉

File: main/lsp/LSPTypechecker.h [hunk 10]
```diff
@@ -124,7 +97,7 @@ class LSPTypechecker final {
     /**
      * Re-typechecks the provided files to re-produce error messages.
      */
-    TypecheckRun retypecheck(std::vector<core::FileRef> frefs, WorkerPool &workers) const;
+    std::vector<std::unique_ptr<core::Error>> retypecheck(std::vector<core::FileRef> frefs, WorkerPool &workers);
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.h:100]: This should still be `const`. The promise is: "`retypecheck` won't mutate any state".

File: main/lsp/LSPTypechecker.h [hunk 11]
```diff
@@ -186,7 +159,7 @@ class LSPTypecheckerDelegate {
     /**
      * Re-typechecks the provided files to re-produce error messages.
      */
-    TypecheckRun retypecheck(std::vector<core::FileRef> frefs) const;
+    std::vector<std::unique_ptr<core::Error>> retypecheck(std::vector<core::FileRef> frefs);
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.h:162]: Ditto.

File: main/pipeline/pipeline.cc [hunk 12]
```diff
@@ -1014,7 +1001,9 @@ ast::ParsedFilesOrCancelled typecheck(unique_ptr<core::GlobalState> &gs, vector<
                                                 make_move_iterator(threadResult.trees.end()));
                     }
                     cfgInferProgress.reportProgress(fileq->doneEstimate());
-                    gs->errorQueue->flushErrors();
+                    for (auto &tree : threadResult.trees) {
```
Comment (jvilk-stripe) [inline main/pipeline/pipeline.cc:1004]: btw: There's one change we need to make here before streaming has user impact.

Right now, the worker threads typecheck files until there are no more, and then send the trees back to the coordinator thread -- so the coordinator thread won't flush those files until the ~end of typechecking. We need to change the loop so that workers periodically send back what they have.

File: core/ErrorFlusher.h [hunk 13]
```diff
@@ -10,9 +10,8 @@ namespace core {
 
 class ErrorFlusher {
 public:
-    virtual void flushErrors(spdlog::logger &logger, std::vector<std::unique_ptr<ErrorQueueMessage>> error) = 0;
-    virtual void flushErrorCount(spdlog::logger &logger, int count) = 0;
-    virtual void flushAutocorrects(const GlobalState &gs, FileSystem &fs) = 0;
+    virtual void flushErrors(spdlog::logger &logger, std::vector<std::unique_ptr<ErrorQueueMessage>> errors,
```
Comment (jvilk-stripe) [inline core/ErrorFlusher.h:13]: (Optional) WDYT about changing the order of arguments to this to put the more globalish things first? Normally, `gs` is near the front of the list, and `file` should probably go before `errors`.

`logger`, `gs`, `file`, `errors`

File: core/ErrorQueue.cc [hunk 14]
```diff
@@ -176,15 +143,20 @@ bool ErrorQueue::isEmpty() {
...
     checkOwned();
-    auto out = drainFlushed();
+    UnorderedMap<core::FileRef, vector<unique_ptr<core::ErrorQueueMessage>>> out;
+
+    core::ErrorQueueMessage msg;
+    for (auto result = queue.try_pop(msg); result.gotItem(); result = queue.try_pop(msg)) {
+        collected[msg.whatFile].emplace_back(make_unique<ErrorQueueMessage>(move(msg)));
+    }
 
     for (auto &part : collected) {
```
Comment (jvilk-stripe) [inline core/ErrorQueue.cc:156]: Can we replace this loop with:

```c++
auto out = move(collected);
collected.clear();
return out;
```

(I forget if `move` on `absl::UnorderedMap` clears the map; I wouldn't be surprised if it didn't.)

File: main/lsp/AutocorrectFlusher.cc [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+void AutocorrectFlusher::flushErrors(spdlog::logger &logger, vector<unique_ptr<core::ErrorQueueMessage>> errors,
+                                     const core::GlobalState &gs, core::FileRef file) {
+    for (auto &error : errors) {
+        if (error->kind == core::ErrorQueueMessage::Kind::Error) {
+            if (error->error->isSilenced) {
+                continue;
+            }
+
+            prodHistogramAdd("error", error->error->what.code, 1);
```
Comment (jvilk-stripe) [inline main/lsp/AutocorrectFlusher.cc:14]: You can remove this call to `prodHistogramAdd`; I think the intent for this metric is to see which errors are getting reported to the user. These errors aren't being reported; we're just mining them for autocorrects.

File: main/lsp/AutocorrectFlusher.cc [hunk 16]
```diff
@@ -0,0 +1,19 @@
+#include "main/lsp/LSPTypechecker.h"
+
+using namespace std;
+namespace sorbet::realmain::lsp {
+
+void AutocorrectFlusher::flushErrors(spdlog::logger &logger, vector<unique_ptr<core::ErrorQueueMessage>> errors,
```
Comment (jvilk-stripe) [inline main/lsp/AutocorrectFlusher.cc:6]: WDYT about renaming this to `ErrorCollector`? It's only used for autocorrect, but it's not actually flushing autocorrects. (It's not actually flushing anything.)

File: main/lsp/AutocorrectFlusher.cc [hunk 17]
```diff
@@ -0,0 +1,19 @@
+#include "main/lsp/LSPTypechecker.h"
```
Comment (jvilk-stripe) [inline main/lsp/AutocorrectFlusher.cc:1]: Why does this depend on `LSPTypechecker`? There are no LSP dependencies here.

File: main/lsp/LSPTypechecker.h [hunk 18]
```diff
@@ -68,29 +51,19 @@ class LSPTypechecker final {
...
     bool initialized = false;
 
-    ErrorReporter errorReporter;
+    std::shared_ptr<ErrorReporter> errorReporter;
 
     /** Conservatively reruns entire pipeline without caching any trees. Returns 'true' if committed, 'false' if
      * canceled. */
     bool runSlowPath(LSPFileUpdates updates, WorkerPool &workers, bool cancelable);
 
     /** Runs incremental typechecking on the provided updates. */
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.h:60]:
```suggestion
    /** Runs incremental typechecking on the provided updates. Returns the final list of files typechecked. */
```

File: main/lsp/LSPTypechecker.h [hunk 19]
```diff
@@ -207,5 +180,15 @@ class LSPTypecheckerDelegate {
     const core::GlobalState &state() const;
 };
 
+class AutocorrectFlusher : public core::ErrorFlusher {
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.h:183]: Oh, what I was saying before is that you don't have to mention `AutocorrectFlusher` in a `.h` file at all. You can just define the entire class in the `.cc` file!

e.g. this is totally fine in a .cc file:

```c++
class AutocorrectFlusher : public core::ErrorFlusher {
  public:
    std::vector<std::unique_ptr<core::Error>> collectedErrors;
    AutocorrectFlusher() = default;
    ~AutocorrectFlusher() = default;
    void flushErrors(spdlog::logger &logger, std::vector<std::unique_ptr<core::ErrorQueueMessage>> errors,
                     const core::GlobalState &gs, core::FileRef file) override {
      // implementation here!
    }
};
```

This way, this class doesn't clutter up the public interface to `LSPTypechecker` (the header file).

File: main/realmain.cc [hunk 20]
```diff
@@ -572,13 +584,18 @@ int realmain(int argc, char *argv[]) {
             }
         }
 
-        gs->errorQueue->flushErrors(true);
+        gs->errorQueue->flushAllErrors(*gs);
 
         if (!opts.noErrorCount) {
-            gs->errorQueue->flushErrorCount();
+            if (gs->errorQueue->nonSilencedErrorCount == 0) {
```
Comment (jvilk-stripe) [inline main/realmain.cc:590]: Optional: Hmm, since `flushAutocorrects` is still in `ErrorFlusherStdout`, maybe we should move this back too.

File: core/ErrorQueue.h [hunk 21]
```diff
@@ -30,24 +29,21 @@ class ErrorQueue {
     std::atomic<int> filesFlushedCount{0};
     bool ignoreFlushes{false};
 
-    ErrorQueue(spdlog::logger &logger, spdlog::logger &tracer);
-    ~ErrorQueue();
+    ErrorQueue(spdlog::logger &logger, spdlog::logger &tracer,
```
Comment (jvilk-stripe) [inline core/ErrorQueue.h:32]: (Optional) I think we can remove `tracer` and use `gs.tracer()` in all of the places it's currently used.

File: core/ErrorFlusherStdout.cc [hunk 22]
```diff
@@ -73,4 +66,12 @@ void ErrorFlusherStdout::flushAutocorrects(const GlobalState &gs, FileSystem &fs
     autocorrects.clear();
 }
 
+void ErrorFlusherStdout::flushErrorCount(spdlog::logger &logger, int count) {
```
Comment (jvilk-stripe) [inline core/ErrorFlusherStdout.cc:69]: (suuuuuuuuuper minor nit) To avoid dirtying the diff you can move this back to its old position ^^

File: main/lsp/LSPTypechecker.cc [hunk 23]
```diff
@@ -22,7 +23,24 @@
 
 namespace sorbet::realmain::lsp {
 using namespace std;
+class ErrorCollector : public core::ErrorFlusher {
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.cc:26]: (Minor nit) Can you put this into the `namespace {` below? That's an anonymous namespace that tells the compiler "this thing is only ever accessed in this file".

Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.cc:26]:
```suggestion
class ErrorCollector final : public core::ErrorFlusher {
```

File: main/realmain.cc [hunk 24]
```diff
@@ -662,7 +674,7 @@ int realmain(int argc, char *argv[]) {
 #endif
     if (!gs || gs->hadCriticalError()) {
         returnCode = 10;
-    } else if (returnCode == 0 && gs->totalErrors() > 0 && !opts.supressNonCriticalErrors) {
+    } else if (returnCode == 0 && gs->errorQueue->nonSilencedErrorCount > 0 && !opts.supressNonCriticalErrors) {
```
Comment (jvilk-stripe) [inline main/realmain.cc:677]: (Nit) Is this change needed? (`gs->totalErrors()` still exists right?)

Comment (vilhena-stripe) [reply main/realmain.cc:677]: oh good call. Not sure why I changed this now
